internal/graph/cached_resolver.go

```diff
@@ -0,0 +1,191 @@
+package graph
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/gob"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultMaxCacheSize     = 10000
+	defaultCacheTTL         = 10 * time.Second
+	defaultResolveNodeLimit = 25
+)
+
+var (
+	checkCacheTotalCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "check_cache_total_count",
+		Help: "The total number of calls to ResolveCheck.",
+	})
+
+	checkCacheHitCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "check_cache_hit_count",
+		Help: "The total number of cache hits for ResolveCheck.",
+	})
+)
+
+// CachedResolveCheckResponse is very similar to ResolveCheckResponse except we
+// do not store the ResolutionData
+type CachedResolveCheckResponse struct {
+	Allowed bool
+}
+
+func (c *CachedResolveCheckResponse) convertToResolveCheckResponse() *ResolveCheckResponse {
+	return &ResolveCheckResponse{
+		Allowed: c.Allowed,
+		ResolutionMetadata: &ResolutionMetadata{
+			Depth:               defaultResolveNodeLimit,
+			DatastoreQueryCount: 0,
```

Comment:
i'm scared that setting `DatastoreQueryCount` to zero will break how we are counting the reads. Shouldn't this be set to `req.ResolutionMetadata.DatastoreQueryCount`?

There is a test for the `DatastoreQueryCount` called `TestCheckDatastoreQueryCount`. Can you create one test that is similar, but which uses this cached resolver? Specifically, I want to make sure that, if one Check call involves several levels in a tree, and the last level is cached, we don't return zero. We should return all the reads we had to do prior to the cached one.